src/ReplaceReferencesCapableTrait.php

```diff
@@ -0,0 +1,99 @@
+<?php
+
+namespace Dhii\Config;
+
+use InvalidArgumentException;
+use Psr\Container\ContainerExceptionInterface;
+use Psr\Container\ContainerInterface;
+use Dhii\Util\String\StringableInterface as Stringable;
+use Psr\Container\NotFoundExceptionInterface;
+
+/**
+ * Trait ReplaceReferencesCapableTrait.
+ *
+ * Adds ability to replace all tokens in the form `${key}` in a string with corresponding values retrieved from a container.
+ *
+ * @since [*next-version*]
+ */
+trait ReplaceReferencesCapableTrait
+{
+    /**
+     * Replaces all tokens in the form `${key}` in a string with corresponding values retrieved from a container.
+     *
+     * @since [*next-version*]
+     *
+     * @param string|Stringable  $input          Input string to find and replace references
+     * @param ContainerInterface $container
+     * @param string|Stringable  $startDelimiter
+     * @param mixed              $default        String to replace reference if key not found in container.
+     * @param string|Stringable  $endDelimiter
+     *
+     * @throws ContainerExceptionInterface
+     * @throws NotFoundExceptionInterface
+     *
+     * @return string The resulting string.
+     */
+    protected function _replaceReferences($input, ContainerInterface $container, $default = null, $startDelimiter = '${', $endDelimiter = '}')
+    {
+        $input = $this->_normalizeString($input);
+
+        $startDelimiter = $this->_normalizeRegexpDelimiter($startDelimiter);
+        $endDelimiter   = $this->_normalizeRegexpDelimiter($endDelimiter);
+
+        $regexp = '/' . $startDelimiter . '(.*?)' . $endDelimiter . '/';
+
+        preg_match_all($regexp, $input, $matches);
+
+        for ($i = 0; $i < count($matches[0]); ++$i) {
+            $token = $matches[0][$i];
+            $key   = $matches[1][$i];
+            try {
+                $value = $container->get($key);
+            } catch (NotFoundExceptionInterface $e) {
+                $value = $default !== null ? $this->_normalizeString($default) : $token;
```

Comment:
No need to normalize `$default` as many times as you have tokens, because it will always be the same.